File: iosApp/iosApp/Features/RecipeDetails/RecipeDetailsSections.swift
import SwiftUI
import Shared

/// Section views ported from recipes/presentation/details/RecipeDetailsSections.kt.

struct IngredientsSection: View {
    let recipe: RecipeDetailUi
    let onAddServing: () -> Void
    let onRemoveServing: () -> Void

    var body: some View {
        SectionCard {
            VStack(alignment: .leading, spacing: 0) {
                Text("Ingredients")
                    .font(SimmerlyFont.headlineSmall)

                HStack {
                    Text(recipe.formattedServings)
                        .font(SimmerlyFont.labelLarge)
                        .foregroundStyle(SimmerlyColor.onSurfaceVariant)
                    Spacer()
                    if recipe.isParsed {
                        HStack(spacing: 4) {
                            ServingStepperButton(
                                systemImage: "minus",
                                label: "Fewer servings",
                                action: onRemoveServing
                            )
                            .disabled(recipe.servings <= 1)
                            ServingStepperButton(
                                systemImage: "plus",
                                label: "More servings",
                                action: onAddServing
                            )
                        }
                    }
                }
                .padding(.top, 12)

                VStack(alignment: .leading, spacing: 0) {
                    ForEach(recipe.ingredients, id: \.referenceId) { ingredient in
                        IngredientEntry(ingredient: ingredient)
                    }
                }
                .padding(.top, 16)

                if !recipe.tools.isEmpty {
                    Text("Tools")
                        .font(SimmerlyFont.headlineSmall)
                        .padding(.top, 24)
                        .padding(.bottom, 8)
                    ForEach(recipe.tools, id: \.id) { tool in
                        Text(tool.name)
                            .font(SimmerlyFont.bodyMedium)
                    }
                }

                Button {
                    // TODO: add to shopping list
                } label: {
                    HStack(spacing: 8) {
                        Image(systemName: "list.bullet.rectangle")
                            .foregroundStyle(SimmerlyColor.onSurfaceVariant)
                        Text("Add to shopping list")
                            .font(SimmerlyFont.titleSmall)
                            .foregroundStyle(SimmerlyColor.onSurface)
                    }
                    .frame(maxWidth: .infinity, minHeight: 40)
                    .overlay(
                        Capsule().strokeBorder(SimmerlyColor.outlineVariant, lineWidth: 1)
                    )
                }
                .buttonStyle(.plain)
                .padding(.top, 16)
            }
        }
    }
}

private struct ServingStepperButton: View {
    let systemImage: String
    let label: String
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            Image(systemName: systemImage)
                .font(.system(size: 15, weight: .medium))
                .frame(width: 32, height: 32)
                .overlay(Circle().strokeBorder(SimmerlyColor.outlineVariant, lineWidth: 1))
        }
        .buttonStyle(.plain)
        .foregroundStyle(SimmerlyColor.onSurfaceVariant)
        .accessibilityLabel(label)
    }
}

/// Two columns: the quantity in a fixed-width gutter so every ingredient name starts on the same
/// vertical line, then the name with its note underneath. Matches IngredientEntry in
/// recipes/presentation/details/RecipeDetailsSections.kt.
private struct IngredientEntry: View {
    let ingredient: IngredientUi

    /// Wider than the design's 78px: that artboard used abbreviated units ("1 tbsp"), but Mealie
    /// spells them out when a unit has useAbbreviation off, and "2 tablespoons" has to fit.
    private static let quantityWidth: CGFloat = 104

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            Divider().overlay(SimmerlyColor.surfaceContainer)
            HStack(alignment: .top, spacing: 10) {
                Text(ingredient.formattedQuantity ?? "")
                    .font(SimmerlyFont.titleSmall)
                    .foregroundStyle(SimmerlyColor.secondary)
                    .frame(width: Self.quantityWidth, alignment: .leading)
                VStack(alignment: .leading, spacing: 0) {
                    Text(ingredient.formattedDisplay)
                        .font(SimmerlyFont.bodyMedium)
                    if let note = ingredient.note, !note.isEmpty {
                        Text(note)
                            .font(SimmerlyFont.bodySmall)
                            .foregroundStyle(SimmerlyPalette.neutral500)
                    }
                }
                Spacer(minLength: 0)
            }
            .padding(.vertical, 9)
        }
    }
}

struct InstructionsSection: View {
    let instructions: [InstructionUi]
    let ingredients: [IngredientUi]

    private var ingredientsById: [String: IngredientUi] {
        Dictionary(uniqueKeysWithValues: ingredients.map { ($0.referenceId, $0) })
    }

    var body: some View {
        SectionCard {
            VStack(alignment: .leading, spacing: 0) {
                Text("Instructions")
                    .font(SimmerlyFont.headlineSmall)

                ForEach(Array(instructions.enumerated()), id: \.element.id) { index, instruction in
                    InstructionEntry(
                        stepNumber: index + 1,
                        instruction: instruction,
                        ingredients: instruction.ingredientIds.compactMap { ingredientsById[$0] }
                    )
                    // Only the first step clears the heading; the rest butt up against the
                    // divider that closes the step above them.
                    .padding(.top, index == 0 ? 12 : 0)
                }
            }
        }
    }
}

private struct InstructionEntry: View {
    let stepNumber: Int
    let instruction: InstructionUi
    let ingredients: [IngredientUi]

    /// Step photos are cropped to a consistent band rather than each taking whatever height its
    /// aspect ratio implies, matching the design.
    private static let imageHeight: CGFloat = 260

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            Divider().overlay(SimmerlyColor.outlineVariant)
            HStack(alignment: .top, spacing: 20) {
                Text("\(stepNumber)")
                    .font(SimmerlyFont.titleSmall)
                    .fontWeight(.bold)
                    .foregroundStyle(SimmerlyColor.secondary)
                    .frame(width: 36, height: 36)
                    .background(SimmerlyColor.primaryContainer, in: Circle())

                VStack(alignment: .leading, spacing: 16) {
                    // Skip the generated "Step N" heading — the badge to the left already says it.
                    if instruction.hasOwnSummary && !instruction.summary.isEmpty {
                        Text(instruction.summary)
                            .font(SimmerlyFont.titleLarge)
                            .foregroundStyle(SimmerlyColor.secondary)
                    }

                    if !ingredients.isEmpty {
                        FlowLayout(spacing: 8) {
                            ForEach(ingredients, id: \.referenceId) { ingredient in
                                IngredientChip(ingredient: ingredient)
                            }
                        }
                    }

                    MarkdownText(markdown: instruction.text)
                        .font(SimmerlyFont.bodyLarge)

                    ForEach(instruction.images, id: \.self) { image in
                        // The image goes in an overlay over a fixed-size spacer rather than
                        // being sized directly: `.aspectRatio(.fill)` reports whatever width it
                        // needs to cover the height, and `.frame(maxWidth: .infinity)` does not
                        // clamp that, so sizing it directly widens the whole card off-screen.
                        Color.clear
                            .frame(maxWidth: .infinity)
                            .frame(height: Self.imageHeight)
                            .overlay {
                                RemoteImage(url: image) {
                                    Rectangle().fill(SimmerlyColor.surfaceContainer)
                                }
                                .aspectRatio(contentMode: .fill)
                            }
                            .clipShape(RoundedRectangle(cornerRadius: 12))
                    }
                }
                Spacer(minLength: 0)
            }
            .padding(.vertical, 20)
        }
    }
}

private struct IngredientChip: View {
    let ingredient: IngredientUi

    var body: some View {
        HStack(spacing: 4) {
            if let quantity = ingredient.formattedQuantity {
                Text(quantity)
                    .fontWeight(.bold)
            }
            Text(ingredient.formattedDisplay)
        }
        .font(SimmerlyFont.bodySmall)
        .foregroundStyle(SimmerlyColor.onTertiaryContainer)
        .padding(.vertical, 4)
        .padding(.horizontal, 8)
        .background(SimmerlyColor.tertiaryContainer, in: RoundedRectangle(cornerRadius: 8))
    }
}

struct NotesSection: View {
    let notes: [Note]

    var body: some View {
        SectionCard {
            VStack(alignment: .leading, spacing: 24) {
                Text("Notes")
                    .font(SimmerlyFont.headlineSmall)

                ForEach(notes, id: \.id) { note in
                    VStack(alignment: .leading, spacing: 6) {
                        if !note.title.isEmpty {
                            Text(note.title)
                                .font(SimmerlyFont.titleLarge)
                                .foregroundStyle(SimmerlyColor.secondary)
                        }
                        Text(note.text)
                            .font(SimmerlyFont.bodyMedium)
                    }
                }
            }
        }
    }
}

struct NutritionSection: View {
    let nutrition: NutritionUi

    var body: some View {
        SectionCard {
            VStack(alignment: .leading, spacing: 0) {
                Text("Nutrition")
                    .font(SimmerlyFont.titleLarge)
                    .padding(.bottom, 12)

                // The label list and its order live in NutritionUi.kt so both platforms render the
                // same facts; this used to be a second copy that had to be kept in step by hand.
                ForEach(nutrition.entries, id: \.label) { entry in
                    VStack(alignment: .leading, spacing: 0) {
                        Divider().overlay(SimmerlyColor.surfaceContainer)
                        HStack {
                            Text(entry.label)
                                .font(SimmerlyFont.bodySmall)
                            Spacer()
                            Text(entry.value)
                                .font(SimmerlyFont.bodySmall)
                                .foregroundStyle(SimmerlyColor.onSurfaceVariant)
                        }
                        .padding(.vertical, 7)
                    }
                }
            }
        }
    }
}

struct RecipeSettingsSheet: View {
    let settings: Settings
    let onSettingChanged: (Settings) -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            Text("Recipe Settings")
                .font(SimmerlyFont.headlineSmall)

            Toggle(
                "Public Recipe",
                isOn: Binding(
                    get: { settings.`public` },
                    set: {
                        onSettingChanged(settings.doCopy(
                            public: $0,
                            showNutrition: settings.showNutrition,
                            showAssets: settings.showAssets,
                            landscapeView: settings.landscapeView,
                            disableComments: settings.disableComments,
                            locked: settings.locked
                        ))
                    }
                )
            )
            Toggle(
                "Show Nutrition Values",
                isOn: Binding(
                    get: { settings.showNutrition },
                    set: {
                        onSettingChanged(settings.doCopy(
                            public: settings.`public`,
                            showNutrition: $0,
                            showAssets: settings.showAssets,
                            landscapeView: settings.landscapeView,
                            disableComments: settings.disableComments,
                            locked: settings.locked
                        ))
                    }
                )
            )
            Toggle(
                "Disable Comments",
                isOn: Binding(
                    get: { settings.disableComments },
                    set: {
                        onSettingChanged(settings.doCopy(
                            public: settings.`public`,
                            showNutrition: settings.showNutrition,
                            showAssets: settings.showAssets,
                            landscapeView: settings.landscapeView,
                            disableComments: $0,
                            locked: settings.locked
                        ))
                    }
                )
            )
            Toggle(
                "Locked",
                isOn: Binding(
                    get: { settings.locked },
                    set: {
                        onSettingChanged(settings.doCopy(
                            public: settings.`public`,
                            showNutrition: settings.showNutrition,
                            showAssets: settings.showAssets,
                            landscapeView: settings.landscapeView,
                            disableComments: settings.disableComments,
                            locked: $0
                        ))
                    }
                )
            )

            Spacer()
        }
        .padding(16)
        .tint(SimmerlyColor.primary)
    }
}

File: sharedUI/src/commonMain/kotlin/dev/juanrincon/simmerly/recipes/presentation/details/RecipeDetailsSections.kt
package dev.juanrincon.simmerly.recipes.presentation.details

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ListAlt
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Remove
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedIconButton
import androidx.compose.material3.ProvideTextStyle
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import coil3.compose.AsyncImage
import com.mohamedrejeb.richeditor.model.rememberRichTextState
import com.mohamedrejeb.richeditor.ui.material3.RichText
import dev.juanrincon.simmerly.core.presentation.ifTrue
import dev.juanrincon.simmerly.recipes.domain.model.Note
import dev.juanrincon.simmerly.recipes.presentation.details.models.IngredientUi
import dev.juanrincon.simmerly.recipes.presentation.details.models.InstructionUi
import dev.juanrincon.simmerly.recipes.presentation.details.models.NutritionUi
import dev.juanrincon.simmerly.recipes.presentation.details.models.RecipeDetailUi
import dev.juanrincon.simmerly.recipes.presentation.details.models.entries
import dev.juanrincon.simmerly.recipes.presentation.shared.IngredientChipRow
import dev.juanrincon.simmerly.theme.Simmerly

/** Inner padding shared by the sections that sit inside a card on every layout. */
private val SECTION_PADDING = 20.dp

/**
 * Width of the quantity column in an ingredient row. Fixed rather than intrinsic so every
 * ingredient name starts on the same vertical line, as the design draws it — a ragged left edge is
 * what a wrap-content quantity column would give on a recipe mixing "24" with "1/2 bunch".
 *
 * Wider than the design's 78px: that artboard used abbreviated units ("1 tbsp"), but Mealie spells
 * them out when a unit has useAbbreviation off, and "2 tablespoons" has to fit on one line.
 */
private val INGREDIENT_QUANTITY_WIDTH = 104.dp

@Composable
internal fun IngredientAndToolView(
    recipe: RecipeDetailUi,
    onAddServingButtonClick: () -> Unit,
    onRemoveServingButtonClick: () -> Unit,
    modifier: Modifier = Modifier
) {
    Column(modifier = modifier.padding(SECTION_PADDING)) {
        Text("Ingredients", style = MaterialTheme.typography.headlineSmall)
        Row(
            modifier = Modifier.fillMaxWidth().padding(top = 12.dp),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                recipe.formattedServings,
                style = MaterialTheme.typography.labelLarge,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
            if (recipe.isParsed) {
                Row(horizontalArrangement = Arrangement.spacedBy(4.dp)) {
                    ServingStepperButton(
                        icon = Icons.Default.Remove,
                        contentDescription = "Fewer servings",
                        onClick = onRemoveServingButtonClick,
                        enabled = recipe.servings > 1
                    )
                    ServingStepperButton(
                        icon = Icons.Default.Add,
                        contentDescription = "More servings",
                        onClick = onAddServingButtonClick
                    )
                }
            }
        }
        Column(modifier = Modifier.fillMaxWidth().padding(top = 16.dp)) {
            recipe.ingredients.forEachIndexed { index, ingredient ->
                IngredientEntry(ingredient, modifier = Modifier.fillMaxWidth(), index != 0)
            }
        }
        if (recipe.tools.isNotEmpty()) {
            Text(
                "Tools",
                style = MaterialTheme.typography.headlineSmall,
                modifier = Modifier.padding(top = 24.dp, bottom = 8.dp)
            )
            recipe.tools.forEach { tool ->
                Text(tool.name, style = MaterialTheme.typography.bodyMedium)
            }
        }
        OutlinedButton(
            onClick = { /* TODO: add to shopping list */ },
            shape = CircleShape,
            modifier = Modifier.fillMaxWidth().padding(top = 16.dp).height(40.dp)
        ) {
            Icon(
                Icons.AutoMirrored.Default.ListAlt,
                contentDescription = null,
                tint = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.size(18.dp)
            )
            Text(
                "Add to shopping list",
                style = MaterialTheme.typography.titleSmall,
                color = MaterialTheme.colorScheme.onSurface,
                modifier = Modifier.padding(start = 8.dp)
            )
        }
    }
}

@Composable
private fun ServingStepperButton(
    icon: ImageVector,
    contentDescription: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    enabled: Boolean = true
) {
    OutlinedIconButton(
        onClick = onClick,
        enabled = enabled,
        shape = CircleShape,
        modifier = modifier.size(32.dp)
    ) {
        Icon(icon, contentDescription = contentDescription, modifier = Modifier.size(18.dp))
    }
}

@Composable
private fun IngredientEntry(
    ingredient: IngredientUi,
    modifier: Modifier = Modifier,
    hasDivider: Boolean = true
) {
    Column(modifier = modifier) {
        if (hasDivider) {
            HorizontalDivider(color = MaterialTheme.colorScheme.surfaceContainer)
        }
        Row(
            modifier = Modifier.fillMaxWidth().padding(vertical = 9.dp),
            horizontalArrangement = Arrangement.spacedBy(10.dp),
            verticalAlignment = Alignment.Top
        ) {
            Text(
                text = ingredient.formattedQuantity.orEmpty(),
                style = MaterialTheme.typography.titleSmall,
                color = MaterialTheme.colorScheme.secondary,
                modifier = Modifier.width(INGREDIENT_QUANTITY_WIDTH)
            )
            Column {
                Text(ingredient.formattedDisplay, style = MaterialTheme.typography.bodyMedium)
                ingredient.note?.let { note ->
                    Text(
                        note,
                        style = MaterialTheme.typography.bodySmall,
                        color = Simmerly.Neutral500
                    )
                }
            }
        }
    }
}

@Composable
internal fun InstructionView(
    instructions: List<InstructionUi>,
    ingredients: List<IngredientUi>,
    modifier: Modifier = Modifier,
    contentPadding: PaddingValues = PaddingValues(SECTION_PADDING)
) {
    val ingredientsById = remember(ingredients) { ingredients.associateBy { it.referenceId } }
    Column(modifier = modifier.padding(contentPadding)) {
        Text("Instructions", style = MaterialTheme.typography.headlineSmall)
        instructions.forEachIndexed { index, instruction ->
            InstructionEntry(
                stepNumber = index + 1,
                instruction = instruction,
                ingredients = instruction.ingredientIds.mapNotNull { ingredientsById[it] },
                // Only the first step clears the heading; the rest butt up against the divider
                // that closes the step above them.
                modifier = Modifier.fillMaxWidth().ifTrue(index == 0) { padding(top = 12.dp) },
                hasDivider = index != 0
            )
        }
    }
}

@Composable
private fun InstructionEntry(
    stepNumber: Int,
    instruction: InstructionUi,
    ingredients: List<IngredientUi>,
    modifier: Modifier = Modifier,
    hasDivider: Boolean = true
) {
    val richTextState = rememberRichTextState()
    LaunchedEffect(instruction.text) {
        richTextState.setMarkdown(instruction.text)
    }
    Column(modifier = modifier) {
        if (hasDivider) {
            HorizontalDivider(color = MaterialTheme.colorScheme.outlineVariant)
        }
        Row(
            modifier = Modifier.fillMaxWidth().padding(vertical = 20.dp),
            horizontalArrangement = Arrangement.spacedBy(20.dp),
            verticalAlignment = Alignment.Top
        ) {
            Box(
                contentAlignment = Alignment.Center,
                modifier = Modifier
                    .size(36.dp)
                    .background(MaterialTheme.colorScheme.primaryContainer, CircleShape)
            ) {
                Text(
                    stepNumber.toString(),
                    style = MaterialTheme.typography.titleSmall,
                    fontWeight = FontWeight.Bold,
                    color = MaterialTheme.colorScheme.secondary
                )
            }
            Column(verticalArrangement = Arrangement.spacedBy(16.dp)) {
                // Skip the generated "Step N" heading — the badge to the left already says it.
                if (instruction.hasOwnSummary && instruction.summary.isNotBlank()) {
                    Text(
                        instruction.summary,
                        style = MaterialTheme.typography.titleLarge,
                        color = MaterialTheme.colorScheme.secondary
                    )
                }
                IngredientChipRow(ingredients, modifier = Modifier.fillMaxWidth())
                // RichText picks its style up from LocalTextStyle; providing it that way keeps
                // us off the experimental `style` parameter while still applying DM Sans and the
                // design's 16/27 step metrics.
                ProvideTextStyle(
                    MaterialTheme.typography.bodyLarge.copy(lineHeight = 27.sp)
                ) {
                    RichText(
                        state = richTextState,
                        modifier = Modifier.widthIn(max = INSTRUCTION_TEXT_MAX_WIDTH)
                    )
                }
                instruction.images.forEach { image ->
                    AsyncImage(
                        model = image,
                        contentDescription = null,
                        contentScale = ContentScale.Crop,
                        modifier = Modifier
                            .widthIn(max = INSTRUCTION_IMAGE_MAX_WIDTH)
                            .height(INSTRUCTION_IMAGE_HEIGHT)
                            .clip(MaterialTheme.shapes.medium)
                    )
                }
            }
        }
    }
}

// Measure caps from the design. Instruction prose is capped so a wide desktop pane keeps a
// readable line length instead of stretching to the full column, and step photos are cropped to a
// consistent band rather than each taking whatever height its aspect ratio implies.
private val INSTRUCTION_TEXT_MAX_WIDTH = 640.dp
private val INSTRUCTION_IMAGE_MAX_WIDTH = 560.dp
private val INSTRUCTION_IMAGE_HEIGHT = 260.dp

@Composable
internal fun NotesView(notes: List<Note>, modifier: Modifier = Modifier) {
    Column(
        modifier = modifier.padding(SECTION_PADDING),
        verticalArrangement = Arrangement.spacedBy(24.dp)
    ) {
        Text("Notes", style = MaterialTheme.typography.headlineSmall)
        notes.forEach { note ->
            NoteEntry(note, modifier = Modifier.fillMaxWidth())
        }
    }
}

@Composable
private fun NoteEntry(note: Note, modifier: Modifier = Modifier) {
    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(6.dp)) {
        if (note.title.isNotEmpty()) {
            Text(
                note.title, style = MaterialTheme.typography.titleLarge,
                color = MaterialTheme.colorScheme.secondary
            )
        }
        Text(note.text, style = MaterialTheme.typography.bodyMedium)
    }
}

@Composable
internal fun NutritionView(nutrition: NutritionUi, modifier: Modifier = Modifier) {
    Column(modifier = modifier.padding(SECTION_PADDING)) {
        Text(
            "Nutrition",
            style = MaterialTheme.typography.titleLarge,
            modifier = Modifier.padding(bottom = 12.dp)
        )
        nutrition.entries.forEachIndexed { index, entry ->
            NutritionEntry(entry.label, entry.value, modifier = Modifier.fillMaxWidth(), index != 0)
        }
    }
}

@Composable
private fun NutritionEntry(
    title: String,
    value: String,
    modifier: Modifier = Modifier,
    hasDivider: Boolean = true
) {
    Column(modifier = modifier) {
        if (hasDivider) {
            HorizontalDivider(color = MaterialTheme.colorScheme.surfaceContainer)
        }
        Row(
            modifier = Modifier.fillMaxWidth().padding(vertical = 7.dp),
            horizontalArrangement = Arrangement.SpaceBetween
        ) {
            Text(title, style = MaterialTheme.typography.bodySmall)
            Text(
                value,
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }
    }
}
